### DISCUSSION TITLE
Next.js: Preload Self-Hosted fonts from node_modules

### DISCUSSION THREAD

**Discussion by alexasselin008** [vercel/next.js#35554] [Help]
In my Next.js application, I want to preload my fonts that are located in the node_modules.

In one of my design system's CSS package, I have  a css file with 
```
/* FONT */
@font-face {
    src: url("TT-Interphases-Var-Roman.woff2") format("woff2-variations");
    font-family: "TT Interphases Variable";
    font-style: normal;
    font-display: swap;
}
```
The font is located in the same folder.

In my _app, I load my main css file
```
import "./main.css";

function App() {
```

And my main.css file imports my design's system css
```
@import "~@orbit-ui/css";
```

The font loads correctly, but I see it flickers when I first load my page. I want to preload that font in order to prevent the flickering.

However I have no way in _document to target that font. The font file is moved to `.next/static/media` with an hash in the name : `TT-Interphases-Var-Roman.4772f5e4.woff2.`

It works fine if i preload my font like this : 
```
<link 
    rel="preload"
    href="_next/static/media/TT-Interphases-Var-Roman.4772f5e4.woff2"
    as="font"
    type="font/woff2"
    crossOrigin=""
/>
```
but if i rebuild my app, the hash in the name will change. 

Is there a better way to reference my font's location?

**Comment by jkjustjoshing** [comment]
Thanks for asking this question. I noticed your [Reddit](https://www.reddit.com/r/nextjs/comments/tmg2ha/nextjs_preload_selfhosted_fonts_from_node_modules/) and [Stack Overflow](https://stackoverflow.com/questions/71592690/next-js-preload-self-hosted-fonts-from-node-modules) posts also have zero responses.

I tried using the `file-loader` webpack plugin to import the font files in my JavaScript, hoping that the resolved file hash would be the same and I could use the imported source destination for my `<link rel="preload">` elements. However, the hash isn't the same as the hash generated by the Next.js webpack step that copies the font files to `.next/static/media`.

At this point I'm VERY close to just copying the fonts to the `/public` folder and copying the CSS into the project, so the font files don't get hashed. But this is a terrible solution.

It seems like Next.js should automatically preload any `@font-face` sources it copies to the output folder. This seems in-line with the performance goals of the project.

**Reply by agusterodin** [reply]
I am in the same boat. Were you able to get file-loader to match the path of the font in .next/_static/bla?

**Comment by alexasselin008** [comment]
I ended up hosting my fonts on a CDN. I then modified my design system to use that CDN, and I then added a link rel="preload" in my application with that same url. 

I saw [the issue ](https://github.com/vercel/next.js/issues/36678) you opened, thank you for that!
